### DISCUSSION TITLE
Default values for TypeVar parameters not binding in classmethod alternate constructors?

### DISCUSSION THREAD

**Discussion by ringohoffman** [microsoft/pyright#7644] [Q&A]
```python
from __future__ import annotations

class MyGeneric[T: int | None]:
    def __init__(self, value: T = None) -> None:  # OK
        self.value = value

    @classmethod
    def build(cls, value: T = None) -> MyGeneric[T]:  # Expression of type "None" cannot be assigned to parameter of type "T@MyGeneric"
        return cls(value)
```

Is there a nuanced reason as to why this is not supported?

**Comment by erictraut** [comment]
Constructors (`__new__` and `__init__` methods) are treated specially because a generic class type does not need to be explicitly specialized prior to the constructor being called; the type arguments can be inferred via the constructor call.

```python
# Type arguments explicitly provided
reveal_type(MyGeneric[Literal[0]](0))  # MyGeneric[Literal[0]]

# Type arguments inferred
reveal_type(MyGeneric(3)) # MyGeneric[int]
```

There's no concept of an "alternative constructor" in the python typing spec. The code in your example above is simply defining a class method. Class methods are bound to a specialized class type. If the class type isn't explicitly specialized (i.e. the type arguments are not explicitly provided), then the type is implicitly specialized using default type parameter values. By default, that means `Any` for TypeVars. 

```python
class MyGeneric[T]:
    @classmethod
    def foo(cls, value: T) -> T:
        ...
reveal_type(MyGeneric[int].foo) # (value: int) -> int
reveal_type(MyGeneric.foo) # (value: Unknown) -> Unknown
```

PEP 696, which will be implemented in Python 3.13, introduces the ability to specify a default value. 

```python
class MyGeneric[T = float]:
    @classmethod
    def foo(cls, value: T) -> T:
        ...
reveal_type(MyGeneric.foo) # (value: float) -> float
```

Within a class method, any parameter or variable that is annotated with type `T` must be assigned a value that is also type `T`. Otherwise it's a type violation.

```python
class MyGeneric[T]:
    @classmethod
    def foo(
        cls,
        value: T = None # Type violation
    ) -> T: 
        x: T = None # Type violation
        return None # Type violation
```

**Reply by ringohoffman** [reply]
Was there an intentional change with how `classmethod`s handled type variables in 1.1.354? Even providing these arguments though non-default variables no longer binds the type variables with classmethods.

```python
from __future__ import annotations

from typing import Generic, TypeVar, Union

T = TypeVar("T", bound=Union[int, None])

class MyGeneric(Generic[T]):
    def __init__(self, value: T) -> None:
        self.value = value

    @classmethod
    def build(cls, value: T) -> MyGeneric[T]:
        return cls(value)

reveal_type(MyGeneric.build(1))
# pyright==1.1.353: Type of "MyGeneric.build(1)" is "MyGeneric[int]"
# pyright>=1.1.354: Type of "MyGeneric.build(1)" is "MyGeneric[Unknown]"
```

**Reply by erictraut** [reply]
Previously, the typing spec was under-specified in this area. The introduction and acceptance of [PEP 696](https://peps.python.org/pep-0696/) clarified how this must work.

I think the typing spec could still be more explicit in this area, so I plan to add a new chapter that further clarifies the behavior when binding a generic class to a method (instance, class or static).

I recently wrote a typing spec chapter specifically focused on [constructor calls](https://discuss.python.org/t/draft-typing-spec-chapter-for-constructors/49744). That chapter is currently under review. Constructor calls are necessarily special-cased in the type system.
